### PR TITLE
Download CSV file for exam's scores

### PR DESCRIPTION
**Pull request by snowmancc**

### CODE REVIEW

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamController.java [hunk 1]
```diff
@@ -161,11 +164,22 @@ public TranscriptView createTranscript(@RequestHeader("Authorization") String au
...
         return tokenService.returnIfAdmin(authorization, token -> {
             ExamTranscriptPresenter presenter = new ExamTranscriptPresenter();
-            calculateExamScoreUseCase.execute(examId, presenter);
+            createExamTranscriptUseCase.execute(examId, presenter);
             return presenter.present();
         });
     }
 
+    @GetMapping(value = "/exams/{examId}/transcript/csv",
+            produces = "application/csv")
```
Comment (timt1028) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamController.java:173]: Maybe merge one Line @GetMapping(value = "/exams/{examId}/transcript/csv", produces = "application/csv")

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java [hunk 2]
```diff
@@ -17,27 +19,38 @@
...
         Exam exam = findExam(examId);
         var examineeRecords = getExamineeRecords(exam);
         presenter.showExam(exam);
         presenter.showRecords(examineeRecords);
+        List<Question> questions = exam.getQuestions();
+        //TODO It should support for getting multiple problems one request
+        List<ProblemView> problems = questions.stream()
+                .map(question -> problemServiceDriver.getProblem(question.getProblemId())
+                        .orElseThrow(() -> notFound(ProblemView.class).id(question.getId().getProblemId())))
+                .collect(toList());
```
Comment (Johnny850807) [inline Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CreateExamTranscriptUseCase.java:52]: Don't throw `NotFoundException`, please just include the problems that exist.

File: Spring-Boot/Spring-Boot-Academy/pom.xml [hunk 3]
```diff
@@ -66,6 +66,11 @@
             <artifactId>h2</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-csv</artifactId>
+            <version>1.8</version>
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Academy/pom.xml:72]: Don't put the version here, please specify the version in the `Dependencies/pom.xml`

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamController.java [hunk 4]
```diff
@@ -161,11 +164,22 @@ public TranscriptView createTranscript(@RequestHeader("Authorization") String au
...
+    public ResponseEntity<InputStreamResource> createCsvFileOfTranscript(@RequestHeader("Authorization") String authorization,
+                                                                         @PathVariable int examId) {
+        return tokenService.returnIfAdmin(authorization, token -> {
+            ExamTranscriptCsvFilePresenter presenter = new ExamTranscriptCsvFilePresenter();
+            createExamTranscriptUseCase.execute(examId, presenter);
+            return ResponseEntityUtils.respondInputStreamResource(presenter.present());
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamController.java:179]: static import

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            }
+            csvPrinter.flush();
+
+            ByteArrayInputStream inStream = new ByteArrayInputStream(out.toByteArray());
+            return new FileResource(EXAM_TRANSCRIPT_FILE_NAME, out.size(), inStream);
+        } catch (IOException e) {
+            throw new RuntimeException("Create csv file of exam transcript error", e);
+        }
+    }
+}
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:102]: Clean Code: Please put the invoked methods below the invoker method

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            for (List<String> record : csvBody) {
+                csvPrinter.printRecord(record);
+            }
+            csvPrinter.flush();
+
+            ByteArrayInputStream inStream = new ByteArrayInputStream(out.toByteArray());
+            return new FileResource(EXAM_TRANSCRIPT_FILE_NAME, out.size(), inStream);
+        } catch (IOException e) {
+            throw new RuntimeException("Create csv file of exam transcript error", e);
+        }
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:100]: @SneakyThrows

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            csvHeader.add(problemView.title);
+        }
+        csvHeader.add(COLUMN_TOTAL_SCORE);
+        return csvHeader.toArray(new String[0]);
+    }
+
+    private List<List<String>> createExamTranscriptCsvBody(List<ExamineeRecord> examineeRecords) {
+        return examineeRecords.stream().map(examineeRecord -> {
+            List<String> row = new ArrayList<>();
+            Student student = examineeRecord.getExaminee();
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:63]: rename `student` to `examinee`

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                    CSVFormat.DEFAULT.withHeader(createExamTranscriptCsvHeader(problems))
+            );
+
+            List<List<String>> csvBody = createExamTranscriptCsvBody(examineeRecords);
+            for (List<String> record : csvBody) {
+                csvPrinter.printRecord(record);
+            }
+            csvPrinter.flush();
+
+            ByteArrayInputStream inStream = new ByteArrayInputStream(out.toByteArray());
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:96]: rename `inStream` to `in`

File: Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamControllerTest.java [hunk 9]
```diff
@@ -628,13 +634,73 @@ void testProduceExamTranscript() {
...
+                .parse(in);
+
+        CSVRecord csvHeader = records.iterator().next();
+        assertEquals(COLUMN_NAME, csvHeader.get(COLUMN_NAME));
+        assertEquals(COLUMN_EMAIL, csvHeader.get(COLUMN_EMAIL));
+        assertEquals(COLUMN_PROBLEM_1_TITLE, csvHeader.get(COLUMN_PROBLEM_1_TITLE));
+        assertEquals(COLUMN_PROBLEM_2_TITLE, csvHeader.get(COLUMN_PROBLEM_2_TITLE));
+        assertEquals(COLUMN_TOTAL_SCORE, csvHeader.get(COLUMN_TOTAL_SCORE));
+
+        Map<String, TranscriptView.ExamineeRecordView> scoreBoard = transcript.scoreBoard;
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamControllerTest.java:674]: var scoreBoard = transcript.scoreBoard;

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

File: Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamControllerTest.java [hunk 10]
```diff
@@ -628,13 +634,73 @@ void testProduceExamTranscript() {
...
+        final String COLUMN_NAME = "Name";
+        final String COLUMN_EMAIL = "Email";
+        final String COLUMN_PROBLEM_1_TITLE = problem.title;
+        final String COLUMN_PROBLEM_2_TITLE = anotherProblem.title;
+        final String COLUMN_TOTAL_SCORE = "Total Score";
+
+        Reader in = new InputStreamReader(new ByteArrayInputStream(response.getContentAsByteArray()));
+        Iterable<CSVRecord> records = CSVFormat.DEFAULT.withHeader
+                (COLUMN_NAME, COLUMN_EMAIL, COLUMN_PROBLEM_1_TITLE, COLUMN_PROBLEM_2_TITLE, COLUMN_TOTAL_SCORE)
+                .parse(in);
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamControllerTest.java:665]: Iterable<CSVRecord> records -> var records

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        csvHeader.add(COLUMN_NAME);
+        csvHeader.add(COLUMN_EMAIL);
+        for (ProblemView problemView : problemViews) {
+            csvHeader.add(problemView.title);
+        }
+        csvHeader.add(COLUMN_TOTAL_SCORE);
+        return csvHeader.toArray(new String[0]);
+    }
+
+    private List<List<String>> createExamTranscriptCsvBody(List<ExamineeRecord> examineeRecords) {
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:60]:
```Java
    private List<List<String>> createExamTranscriptCsvBody(List<ExamineeRecord> examineeRecords) {
        return mapToList(examineeRecords, this::examTranscriptCsvRows);
    }

    private List<String> examTranscriptCsvRows(ExamineeRecord examineeRecord) {
        return flatMapToList(of(examineeCsvBodyRows(examineeRecord),
                questionScoreCsvBodyRows(problems, examineeRecord)), Collection::stream);
    }

    private List<String> examineeCsvBodyRows(ExamineeRecord examineeRecord) {
        var examinee = examineeRecord.getExaminee();
        return of(examinee.getName(), examinee.getEmail());
    }

    private List<String> questionScoreCsvBodyRows(Collection<ProblemView> problems,
                                                  ExamineeRecord examineeRecord) {
        List<String> scoreRows = new ArrayList<>(1 + problems.size());
        var questionScores = questionScores(examineeRecord);
        int totalScore = 0;
        for (var problem : problems) {
            int score = questionScores.get(problem.id);
            scoreRows.add(String.valueOf(score));
            totalScore += score;
        }
        scoreRows.add(String.valueOf(totalScore));
        return scoreRows;
    }

    private GetById<Integer, Integer> questionScores(ExamineeRecord examineeRecord) {
        var questionRecords = examineeRecord.getQuestionRecords();
        var questionScores = toMap(questionRecords,
                questionRecord -> questionRecord.getQuestion().getProblemId(),
                CreateExamTranscriptUseCase.QuestionRecord::calculateScore);
        return questionScores::get;
    }

    public FileResource present() {
        try (var out = new ByteArrayOutputStream()) {
            var csvPrinter = new CSVPrinter(new PrintWriter(out, false, UTF_8),
                    CSVFormat.DEFAULT.withHeader(createExamTranscriptCsvHeader(problems)));

            var csvBody = createExamTranscriptCsvBody(examineeRecords);
            for (List<String> record : csvBody) {
                csvPrinter.printRecord(record);
            }
            csvPrinter.flush();

            var in = new ByteArrayInputStream(out.toByteArray());
            return new FileResource(EXAM_TRANSCRIPT_FILE_NAME, out.size(), in);
        } catch (IOException e) {
            throw new RuntimeException("Create csv file of exam transcript error", e);
        }
    }
```

**Review by Wally5077 — CHANGES_REQUESTED** [review]

File: API/Problem-API/src/main/java/tw/waterball/judgegirl/problemapi/clients/RestProblemApiClient.java [hunk 12]
```diff
@@ -68,4 +73,18 @@ public FileResource downloadTestCaseIOs(int problemId, String testcaseId) throws
         }
     }
 
+    @Override
+    public List<ProblemView> getProblemsByIds(List<Integer> ids) {
+        try {
+            String idsSplitByComma = ids.stream().map(String::valueOf).collect(joining(","));
```
Comment (Wally5077) [inline API/Problem-API/src/main/java/tw/waterball/judgegirl/problemapi/clients/RestProblemApiClient.java:79]: 這一段我認為可以放在 StreamUtils ，多開 join 方法寫這段邏輯
ids.stream().map(String::valueOf).collect(joining(","));

File: API/Problem-API/src/main/java/tw/waterball/judgegirl/problemapi/clients/RestProblemApiClient.java [hunk 13]
```diff
@@ -68,4 +73,18 @@ public FileResource downloadTestCaseIOs(int problemId, String testcaseId) throws
...
+            String url = parsePath(API_PREFIX + "?ids=", idsSplitByComma);
```
Comment (Wally5077) [inline API/Problem-API/src/main/java/tw/waterball/judgegirl/problemapi/clients/RestProblemApiClient.java:80]: 多載 parsePath(url, map)，使它支援 query parameters

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        return scoreRows;
+    }
+
+    private GetById<Integer, Integer> questionScores(ExamineeRecord examineeRecord) {
+        var questionRecords = examineeRecord.getQuestionRecords();
+        var idToQuestionScores = toMap(questionRecords,
+                questionRecord -> questionRecord.getQuestion().getProblemId(),
+                CreateExamTranscriptUseCase.QuestionRecord::calculateScore);
+        return idToQuestionScores::get;
+    }
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/presenters/ExamTranscriptCsvFilePresenter.java:110]: 最後還是改用 Map 好了，這樣可以直接使用 Map.getOrDefault(key, defaultValue)，確保一定有資料

```Java
private List<String> questionScoreCsvBodyRows(Collection<ProblemView> problems,
                                                  ExamineeRecord examineeRecord) {
        List<String> scoreRows = new ArrayList<>(1 + problems.size());
        var questionScores = questionScores(examineeRecord);
        int totalScore = 0;
        for (var problem : problems) {
            int score = questionScores.getOrDefault(problem.id, 0);
            scoreRows.add(String.valueOf(score));
            totalScore += score;
        }
        scoreRows.add(String.valueOf(totalScore));
        return scoreRows;
    }

    private Map<Integer, Integer> questionScores(ExamineeRecord examineeRecord) {
        var questionRecords = examineeRecord.getQuestionRecords();
        return toMap(questionRecords,
                questionRecord -> questionRecord.getQuestion().getProblemId(),
                CreateExamTranscriptUseCase.QuestionRecord::calculateScore);
    }
```

**Review by Wally5077 — APPROVED** [review]
LGTM

**Review by timt1028 — COMMENTED** [review]
LGTM

**Review by timt1028 — APPROVED** [review]
LGTM

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Wait for CI's pass

**Pull request merged**